File: fixtures/coding/maintainability-mini-service-go/existing_customers_test.go
package miniservice

import (
	"reflect"
	"testing"
)

func adminUser() *User {
	return &User{ID: "admin-1", Role: "admin"}
}

func TestExistingCreateAndListCustomerFlowStillWorks(t *testing.T) {
	app := BuildApp()

	create := app.Handle(Request{
		Method: "POST",
		Path:   "/customers",
		JSON: map[string]any{
			"name":  "Ada Lovelace",
			"email": "Ada@Example.COM",
			"plan":  "pro",
			"tags":  []any{"VIP"},
		},
		User: adminUser(),
	})
	if create.StatusCode != 201 {
		t.Fatalf("status = %d, want 201", create.StatusCode)
	}
	customer := create.Body["customer"].(map[string]any)
	if customer["email"] != "ada@example.com" {
		t.Fatalf("email = %v", customer["email"])
	}
	if !reflect.DeepEqual(customer["tags"], []string{"vip"}) {
		t.Fatalf("tags = %#v", customer["tags"])
	}

	listed := app.Handle(Request{Method: "GET", Path: "/customers", User: adminUser()})
	if listed.StatusCode != 200 {
		t.Fatalf("list status = %d", listed.StatusCode)
	}
	customers := listed.Body["customers"].([]map[string]any)
	if got := customers[0]["email"]; got != "ada@example.com" {
		t.Fatalf("listed email = %v", got)
	}
}

func TestExistingCreateCustomerRejectsInvalidPayload(t *testing.T) {
	app := BuildApp()
	response := app.Handle(Request{Method: "POST", Path: "/customers", JSON: map[string]any{"name": "", "email": "bad", "plan": "gold"}, User: adminUser()})
	if response.StatusCode != 400 {
		t.Fatalf("status = %d, want 400", response.StatusCode)
	}
	errors := response.Body["errors"].([]string)
	assertContains(t, errors, "name is required")
	assertContains(t, errors, "email must be valid")
	assertContains(t, errors, "plan is invalid")
}

func TestExistingCreateCustomerRequiresPermission(t *testing.T) {
	app := BuildApp()
	response := app.Handle(Request{Method: "POST", Path: "/customers", JSON: map[string]any{"name": "Grace", "email": "grace@example.com"}, User: &User{ID: "viewer"}})
	if response.StatusCode != 403 {
		t.Fatalf("status = %d, want 403", response.StatusCode)
	}
	if len(app.Repository.ListCustomers()) != 0 {
		t.Fatalf("repository mutated")
	}
}

File: fixtures/coding/maintainability-mini-service-ts/tests/existing-customers.test.ts
import { describe, expect, test } from "vitest";
import { buildApp } from "../src/container";
import type { User } from "../src/models";

const admin = (): User => ({ id: "admin-1", role: "admin" });

describe("existing customer routes", () => {
  test("create and list customer flow still works", () => {
    const app = buildApp();

    const create = app.handle({
      method: "POST",
      path: "/customers",
      json: { name: "Ada Lovelace", email: "Ada@Example.COM", plan: "pro", tags: ["VIP"] },
      user: admin(),
    });

    expect(create.statusCode).toBe(201);
    const customer = create.body.customer as { email: string; tags: string[] };
    expect(customer.email).toBe("ada@example.com");
    expect(customer.tags).toEqual(["vip"]);

    const listed = app.handle({ method: "GET", path: "/customers", user: admin() });
    expect(listed.statusCode).toBe(200);
    expect((listed.body.customers as Array<{ email: string }>).map((item) => item.email)).toEqual(["ada@example.com"]);
  });

  test("create customer rejects invalid payload", () => {
    const app = buildApp();
    const response = app.handle({ method: "POST", path: "/customers", json: { name: "", email: "bad", plan: "gold" }, user: admin() });

    expect(response.statusCode).toBe(400);
    expect(response.body.errors).toContain("name is required");
    expect(response.body.errors).toContain("email must be valid");
    expect(response.body.errors).toContain("plan is invalid");
  });

  test("create customer requires permission", () => {
    const app = buildApp();
    const response = app.handle({ method: "POST", path: "/customers", json: { name: "Grace", email: "grace@example.com" }, user: { id: "viewer" } });

    expect(response.statusCode).toBe(403);
    expect(app.repository.listCustomers()).toEqual([]);
  });
});
